src/client.rs

```diff
@@ -0,0 +1,162 @@
+use cosmian_crypto_core::{
+    reexport::rand_core::SeedableRng, Aes256Gcm, CryptoCoreError, CsRng, Dem,
+    FixedSizeCBytes, Instantiable, Nonce, RandomFixedSizeCBytes, SymmetricKey,
+};
+use rand::{Rng, RngCore};
+use std::io::{Error, ErrorKind};
+
+use crate::{btree::DataItem, oram::BUCKET_SIZE};
+
+pub struct ClientORAM {
+    pub stash: Vec<DataItem>,
+    csprng: CsRng,
+    key: SymmetricKey<{ Aes256Gcm::KEY_LENGTH }>,
+}
+
+impl ClientORAM {
+    pub fn new() -> ClientORAM {
+        let mut csprng = CsRng::from_entropy();
+        let key = SymmetricKey::new(&mut csprng);
+
+        ClientORAM {
+            // Empty stash at initialization.
+            stash: Vec::new(),
+            csprng,
+            key,
+        }
+    }
+
+    pub fn gen_key(&mut self) {
+        self.key = SymmetricKey::new(&mut self.csprng);
+    }
+
+    pub fn generate_dummies(
+        &mut self,
+        nb_dummies: usize,
+        block_size: usize,
+    ) -> Result<Vec<DataItem>, Error> {
+        if !(nb_dummies + 1).is_power_of_two() {
+            return Err(Error::new(
+                ErrorKind::InvalidInput,
+                "Number of blocks shall be power of 2 minus one".to_string(),
+            ));
+        }
+
+        // Number of leaves is 2^(l-1) if number of blocks is 2^l - 1.
+        let nb_leaves = (nb_dummies + 1) / 2;
+        let mut dummies = Vec::new();
+
+        let cipher = Aes256Gcm::new(&self.key);
+
+        // FIXME - encrypt fixed dummy value instead of encrypting randoms.
+        for _ in 0..nb_dummies * BUCKET_SIZE {
+            // Generate new random dummy data.
+            let mut dummy_data = vec![0; block_size];
+            self.csprng.fill_bytes(&mut dummy_data);
+
+            // Generate new nonce for encryption.
+            let nonce = Nonce::new(&mut self.csprng);
+
+            // Encrypt dummies to provide correct MAC for later decryption.
+            let ciphertext_res =
+                cipher.encrypt(&nonce, &dummy_data, Option::None);
+
+            if ciphertext_res.is_err() {
+                return Err(Error::new(
+                    ErrorKind::Interrupted,
+                    Result::unwrap_err(ciphertext_res).to_string(),
+                ));
+            }
```

Comment:
The comment is wrong imho.

You can use a `map_err` to convert the error and `?` to unwrap it.
```suggestion
            let encrypted_data = cipher.encrypt(&nonce, &dummy_data, None).map_err(|e|
                Error::new(ErrorKind::Interrupted, e.to_string()))?;
```

Reply:
How is this comment wrong ? If you try to GCM decrypt random data, the concatenated MAC will not verify random data as if it had been encrypted ?

Reply:
Concerning errors, I didn't know you could handle them like that! So much better :)

Reply:
Done in 52dd2d82e4d6d48398c8fb028095f40120e6de74

Tho, I'm curious to know your position on the comment being inaccurate.

Reply:
It is actually misleading rather than wrong. You encrypt 0s in order to have a seemingly valid value. The server won't be able to check the MAC. You could even use a wrong mac to discriminate dummies from valid encrypted values.

Or maybe I am missing something here.

Reply:
First thing is we should not encrypt a vector of 0s since it is used internally in GCM to compute the authentication key that holds for the MAC

Discriminating dummies should could be done by checking if the MAC is invalid but if a malicious third party interferes with the communications and changes the ciphertexts, we wouldn't be able to tell.

How to discriminate them is matter of discussion more than code reviewing I think.

Reply:
I agree for your second point.

For the first, I don't know that. I will check. Anyway you still can encrypt 1s and my point still stands. Anyway, this is setup and generating too much random shouldn't be a problem so do as you please. Encrypting random just sounds redundant to me.